Designing Metrics Blog (#2142)
fishtown-analytics/docs.getdbt.com: website/blog/2022-10-18-how-to-design-metrics.md

---

> title: "How to design metrics: deciding what to model in the modeling layer vs the semantic layer"

Reviewer: Love the way we incorporated the thesis in the title! This is a bit wordy - perhaps something along the lines of "How to design metrics: deciding which logic lives modeling layer vs the semantic layer"

> description: "placeholder"

Reviewer: todo: get pithy description of this blog

Author: Don't love it but added something

slug: how-to-design-metrics

authors: [callum_mccann]

tags: [dbt product updates]
hide_table_of_contents: false

date: 2022-10-18
is_featured: true
---

---

> **IMPORTANT:** This document serves as the temporary location for information on how to design and structure your metrics. It is our intention to take this content and turn it into a Guide, similar to [How we structure our dbt projects](https://docs.getdbt.com/guides/best-practices/how-we-structure/1-guide-overview) , but we feel that codifying information in a Guide first requires that Metrics be rigorously tested by the community so that best practices can arise. This document contains our early attempts to create best practices. In other words, read these as suggestions for a new paradigm and share in the community where they do (or don’t) match your experiences! More information on where to do this can be found at the end.

Reviewer 2: switch active voice
```suggestion
**IMPORTANT:** This document serves as the temporary location for information on how to design and structure your metrics. It is our intention to take this content and turn it into a Guide, like [How we structure our dbt projects](https://docs.getdbt.com/guides/best-practices/how-we-structure/1-guide-overview), but we feel that codifying information in a Guide first requires that Metrics be rigorously tested by the community so that best practices can arise. This document contains our early attempts to create best practices. In other words, read these as suggestions for a new paradigm and share in the community where they do (or don’t) match your experiences! You can find more information on where to do this at the end.
```

---

## The power of a semantic layer on top of a mature data modeling framework

> As a longtime dbt Community member, I knew I had to get involved when I first saw the dbt Semantic Layer. It gave me a vision of a world where metrics and business logic were unified across an entire organization; a world where the data team was no longer bound to a single consuming experience and could enable their stakeholders in dozens of different ways. To me, it felt like the opportunity to contribute to the next step of what dbt could become.

Reviewer: Do you remember the exact moment? Drew's talk at Coalesce? something else?

Author: I mean its a bit of a historical distortion but if I had to pin down a single moment it was actually probably the "dbt should know about metrics" PR that opened my eyes to dbt being more than it was.

> In past roles, I’ve been referred to as the `dbt zealot` and I’ll own that title! But it’s not a surprise - dbt was built to serve analytics engineers exactly like me. It gave us flexibility and power to serve our organizations. But it always felt like dbt best served **us.**

Reviewer: ```SUGGESTION
In past roles, I’ve been referred to as the `dbt zealot` and I’ll gladly own that title as dbt was built to allow data practitioners like me to greatly expand the power of our work by writing code based off of software engineering principles. It gave us flexibility and power to serve our organizations. But I always wondered if there were more folks who could directly benefit from interacting with dbt.

Author: made some changes but kept the first part because I like the flow of a hard cut after "I'll own that title!"

> The Semantic Layer takes this to the next level **by coupling dbt’s mature data modeling framework with semantic definitions.** The result is a first of its kind data experience that serves both the analytics engineers and stakeholders. Metrics are the first step towards this vision, allowing users to version control and centrally define their key business metrics in a single repo **while also** serving them to the entire business.

Reviewer: ```SUGGESTION 
The Semantic Layer takes expands the reach of dbt **by coupling dbt’s mature data modeling framework with semantic definitions.** The result is a first of its kind data experience that serves both the data practitioners writing your analytics code and stakeholders who depend on it. Metrics are the first step towards this vision, allowing users to version control and centrally define their key business metrics in a single repo **while also** serving them to the entire business.

> However, this is still a relatively new part of the dbt tool box and you probably have a lot of questions on how **exactly** you can do that. Contained within this blog are our early recommendations for how to design and structure metrics, along with examples so that you can apply the same principles to your use cases. They were developed by combining the overall philosophy of dbt with our hands on learning gathered during the beta period and internal testing.

Reviewer 2: ```suggestion
However, this is still a relatively new part of the dbt toolbox and you probably have a lot of questions on how **exactly** you can do that. This blog contains our early recommendations for how to design and structure metrics, along with examples so that you can apply the same principles to your use cases. We developed these recommendations by combining the overall philosophy of dbt, with our hands-on learning gathered during the beta period and internal testing.
```

## How to design your metrics

> **Pre-reading:** We recommend reading through the [metrics documentation](https://docs.getdbt.com/docs/building-a-dbt-project/metrics), which contains a table of all the required/optional properties. 

Reviewer 2: ```suggestion
**Pre-reading:** We recommend reading through the [metrics documentation](/docs/building-a-dbt-project/metrics), which contains a table of all the required/optional properties. 
```

### When to put business logic in the semantic layer vs the modeling layer

> Our instinct when designing metrics might be to encode as much as possible into the semantic layer. An example of this is case statements - the analytics engineer’s gut instinct might be to mimic tools of the past and provide complicated case statements for the metric `expression` property to try and capture the nuance of how it should be calculated. 

Reviewer: As much what - information?

> With the coupling of dbt’s modeling framework, this isn’t necessarily the right answer. In situations with more complicated logic, it might make more sense to keep this logic contained within the model and then reference it in the semantic layer.

Reviewer: ```SUGGESTION 
But remember - you always have the option of performing this logic _in the modeling layer_. This is the key difference between dbt and other semantic layer offerings - by sitting the semantic layer atop a mature transformation layer, you always have the option to configure and optimize your logic within your models and then _define semantic components with intentionality_.

> Getting the balance right has been a learning experience for our team as well. It’s easy to fall into comfortable patterns that might not be the best solution - it’s important for us to take a step back and think from first principles about the line of separation between what should live in a model and what should live in a metric. To explore this, we’ll walk through two examples of handling this divide.

Reviewer: ```SUGGESTION 
Getting the balance  just right is a learning experience and developing community best practices and standards will take time, which is why it’s important for us to think from first principles. What should really be our goal when determining whether logic lives in a model or a metrics? 

To explore this question and begin to develop an intuition, we’ll walk through two examples of handling this divide.

> ### Basic example - designing your metrics

Reviewer: I think we can tweak the headings here to make the sectioning clearer by tweaking our H2s and H3s.

## The power of a semantic layer on top of a mature data modeling framework
### 
When to put business logic in the semantic layer vs the modeling layer

## Basic example - revenue
### Designing your metric
### logic in the modeling layer vs the semantic layer
### choosing which dimensions to include on the metric

## advanced example - NPS
### designing a complex metric
### option 1 - keeping everything in the semantic layer
### option 2 - shift logic left into the modelling layer

## How to structure your metrics
### Folder structure 
### File structure 

## A call to action

In this example, we’ll cover the basics of defining a metric and a fairly straightforward example of where users can draw the line between the semantic layer and the modeling layer. You should finish this section with a better understanding of dbt metrics and its relationship to the modeling layer.

> We begin by taking a look at our end-state marts model called `order_events` that looks something like the below table, but on the order of millions of rows instead of five. Our finance team uses the below model to better understand revenue but inconsistencies in how its reported have led to requests that the data team centralize the definition in the dbt repo.

Reviewer: Can we give a sense on _why_ we begin with this mart. Something like "In the past the mart was often your end stage layer before data was consumed in another tool or system. Now, the mart is our springboard for the creation of our metric"

Author: Added

        - In a SaaS business though, it might make more sense - especially with usage based pricing. The Customer Success team might be interested in tracking the revenue of certain customers and ensuring that they remain consistent.

To quote Cameron Afzal, Product Manager of the dbt Semantic Layer:

> Thoughtful curation of dimensions is essential for three main reasons:
- ****Relevance:**** Analysts must include the dimensions most relevant to answering the question.
- ****Trust****: Curating high-quality dimensions with little to no known errors helps ensure trust in analysis results and the decisions that follow.
- ****Efficiency****: Curation provides a faster path to high-quality analysis results.
> 

To put it another way, **metrics are most useful when every dimension provided can help provide answers to the business.** 

> ### Advanced example - how to design complicated metrics

Reviewer: We might want to call this "complex" rather than "complicated" throughout

Reviewer 2: that's a great suggestion! complex or advanced is a good partner with 'basic'

> ### Advanced example - how to design complicated metrics

Reviewer 2: ```suggestion
### Advanced example - how to design complex metrics
```

> Now let’s look at a more complicated example of a metric that is calculated from components that could be metrics unto themselves. The metric in question is Net Promoter score, which is used by the dbt Labs internal analytics team to understand the experience that users are having on dbt Cloud.

Reviewer: ```SUGGESTION
Now let’s look at a more complex example of a metric  - one that is built from components that could theoretically themselves be metrics. The metric in question is Net Promoter score, which is used by the dbt Labs internal analytics team to understand the experience that users are having on dbt Cloud.

Reviewer: Also this may or may not be too much of a throwback but 
![image](https://user-images.githubusercontent.com/8147311/194768432-66ef0941-0ee7-4751-8782-f3f575485c57.png)

For those of you who are unfamiliar with the industry metric of Net Promoter Score, here is a [great article from the folks over at Delighted on how it is calculated.](https://delighted.com/net-promoter-score) The short version of it is `the percentage of promoters - the percentage of detractors`. 

---

Here at dbt Labs we provide users with short surveys where they can provide feedback (as well as in a few other locations). The data is collected from those surveys is used to calculate our NPS Score, which helps us understand user sentiment over time.

> Given that these surveys come from a few different sources, there is a substantial amount of upstream modeling performed in order to unify them in a single model, but the end result is something that looks similar to the table below:

Reviewer 2: ```suggestion
Given that these surveys come from a few different sources, there is a large amount of upstream modeling performed in order to unify them in a single model, but the end result is something that looks like the table below:
```

| feedback_date | unique_id | feedback_source | user_type | account_plan | score | nps_category |
| --- | --- | --- | --- | --- | --- | --- |
| 2022-10-01 | 1 | nps_tool_1 | developer | team | 5 | detractor |
| 2022-10-01 | 2 | nps_tool_2 | read_only | developer | 8 | promoter |
| 2022-10-02 | 3 | nps_tool_1 | developer | enterprise | 10 | promoter |
| 2022-10-02 | 4 | nps_tool_1 | developer | developer | 7 | passive |
| 2022-10-02 | 5 | nps_tool_2 | developer | team | 9 | promoter |
| 2022-10-03 | 6 | nps_tool_1 | developer | enterprise | 7 | passive |

> The dbt Internal Analytics team ([long may they reign](https://www.linkedin.com/feed/update/urn:li:activity:6962884130569080833/)) took this data and decided to build the NPS Score metric into our repo so that it could be surfaced to stakeholders in multiple tools. This process is where we began to form our opinions on what should live in the modeling layer vs semantic layer.

Reviewer: I know I have said that we caveated in too many places, but I think that a caveat might be useful here along the lines of. "This process is where we began to form our early opinions on what should live in the modeling layer vs semantic layer - but these are sure to progress as we add in more and more real world use cases."

The underlying principle of why this works is based on the fact that averages divide the sum of the values in the set by their number. In more dbt friendly terms, what it really means is that average is creating the following equation: `sum(value)/count(*)`. In the first example implementation, we were doing roughly the same thing with multiple metric definitions - the only difference was our numerator was a count that assigned each row a value of 1. So if we duplicate that logic and give each row a value of 1 then we can create far fewer metrics.

But that only gets us to the `promoter_pct` and `detractor_pct` metrics. In order to combine these both into a single metric definition, we needed to change the value that we assign. Given that the total range of values that the metric could output is -100 (all detractors) to 100 (all promoters) we can assign each of those categories that peak value, along with 0 for passives. This means that when the numerator is aggregated, it nets out promoters against detractors just like the documented equation does `promoter score - detractor score` . 

</details>

**Is this what I should do?**

> It depends! There will be times when it might be better to have logic stored in the modeling layer and there will be times when it might be better to have logic stored in the semantic layer. The needs of your business stakeholders should drive your decision on where to keep this logic.

Reviewer: 1. Should we link this off of the "it depends" https://twitter.com/SBinLondon/status/1413113782214266885
2. Can we give some flavor to how folks might choose between the "it depends"

Author: 1. Yea!
2. I honestly don't know if we/me yet have opinions on when to do each option. The best short hand I've gotten has been "Put it all in the semantic layer if you want to analyze/use the individual components as well"

Author: Added a comment here

## How to structure your metrics

> Now that we’ve designed our metrics, we move on to structuring them within our project. There are different ways to organize your metrics definitions and each comes with pros and cons.

Reviewer: ```SUGGESTION 
Now that we’ve designed our metrics, let's move on to structuring them within our project. We'll examine the different ways to organize metrics  and take a look at the pros and cons of several strategies.

The single-file-code-owner method would look like this:

```yaml
models: 
	metrics:
		marts:
			- revenue.yml	
			- average_order_value.yml	
			- some_other_metric_name.yml
```

### Folder and file structure is a preference, not a hard rule

In the end, all of the structuring information above is just a recommendation. Your project probably has a defined convention in how nodes are organized, whether or not it follows dbt best practices, and you should continue to follow your own organizational practices. That said, we do recommend that metrics should be separate from model yml files. The reason?

**Metrics are important business objects unto themselves and should live separate from the model definitions.**

## Looking To The Future

This is just the beginning of dbt Metrics and the Semantic Layer. We have a number of exciting ideas for expanding capabilities that we plan to begin work on in the coming months. However, we can’t do that without you. 

> If you’ve gotten this far on a technical document on how to design and structure metrics then I’m going to assume you’re a member of the dbt community. Y’all are going to be our guiding light to help us make sure that all the functionality we add helps **you** serve the needs of your business. So if your developer experiences with the Semantic Layer match what we’ve written in this post, and especially if they don’t, I welcome [comments and feedback in this Discourse Discussion](https://discourse.getdbt.com/t/how-to-design-and-structure-metrics-the-discussion/4965)! Additionally, I recommend that you join us over at #dbt-core-metrics on the dbt Slack where I’ll be posting updates, answering questions, discussing usage, and hopefully responding with the best emojis.

Reviewer: ```SUGGESTION 
This semantic layer is a fundamental change to what it means to interact with dbt and ultimately most of the best practices will come from the dbt Community - folks like you. It does not matter if you consider yourself an "expert" on this - we want to talk to you and hear how you are using or would like to use metrics and the semantic layer.  Y’all are going to be our guiding light to help us make sure that all the functionality we add helps **you** serve the needs of your business. 

If your experience with the Semantic Layer match what we’ve written in this post, and especially if they don’t, please share [comments and feedback in this Discourse Discussion](https://discourse.getdbt.com/t/how-to-design-and-structure-metrics-the-discussion/4965)! 

Additionally, I would invite you to join us over at #dbt-core-metrics on the dbt Slack where we’ll be posting updates, answering questions, discussing usage, and hopefully responding with the best emojis.
